packages/parser/src/lezer-to-dast/lezer-to-dast.ts

```diff
@@ -0,0 +1,372 @@
+import { SyntaxNode, TreeBuffer } from "@lezer/common";
+import {
+    CloseTag,
+    MissingCloseTag,
+    OpenTag,
+    TagName,
+} from "../generated-assets/lezer-doenet.terms";
+import { parser } from "../generated-assets/lezer-doenet";
+import {
+    DastAttribute,
+    DastDoctype,
+    DastElementContent,
+    DastFunctionMacro,
+    DastMacro,
+    DastNodes,
+    DastRoot,
+    DastRootContent,
+    DastText,
+    LezerSyntaxNodeName,
+} from "../types";
+import { createErrorNode } from "./create-error-node";
+import {
+    attributeValueText,
+    createOffsetToPositionMap,
+    entityToString,
+    extractContent,
+    extractDoctypeInfo,
+    findFirstErrorInChild,
+    getLezerChildren,
+    lezerNodeToPosition,
+    textNodeToText,
+    updateNodePositionData,
+} from "./lezer-to-dast-utils";
+import { parseMacros } from "../macros";
+import { gobbleFunctionArguments } from "./gobble-function-arguments";
+
+/**
+ * Create a lezer `SyntaxNode` from a string. This can be passed
+ * to `lezerToDast` to create a DAST tree.
+ */
+export function stringToLezer(source: string): SyntaxNode {
+    const tree = parser.parse(source);
+    return tree.topNode;
+}
+
+/**
+ * Convert a lezer `SyntaxNode` into a DAST tree.
+ */
+export function lezerToDast(
+    node: SyntaxNode | string,
+    source?: string,
+): DastRoot {
+    if (typeof node === "string") {
+        const tree = parser.parse(node);
+        source = node;
+        return _lezerToDast(tree.topNode, source);
+    }
+    if (source == null) {
+        throw new Error(
+            `If you provide a SyntaxNode, you must also provide the source string`,
+        );
+    }
+    return _lezerToDast(node, source);
+}
+
+export function _lezerToDast(node: SyntaxNode, source: string): DastRoot {
+    if (typeof node === "string") {
```

Comment:
Delete this check since already know node is a SyntaxNode? Certainly true if `_lezerToDast` is not exported outside this file.